File: 0601-0700/0668-Kth Smallest Number in Multiplication Table/0668-Kth Smallest Number in Multiplication Table.cpp
class Solution {
public:
    int findKthNumber(int m, int n, int k) {
        if (m > n) {
            swap(m, n);
        }
        int start = 1, end = m * n;
        while (start < end) {
            int mid = start + (end - start) / 2;
            int count = 0;
            for (int i = 1; i <= m; ++i) {
                count += min(mid / i, n);
            }
            
            if (count < k) {
                start = mid + 1;
            }
            else {
                end = mid;
            }
        }
        
        return start;
    }
};

File: 0601-0700/0668-Kth Smallest Number in Multiplication Table/0668-Kth Smallest Number in Multiplication Table.py
class Solution:
    def findKthNumber(self, m: int, n: int, k: int) -> int:
        if m > n:
            m, n = n, m
        low = 1
        high = m * n
        while low < high:
            mid = (low + high) // 2
            count = 0
            for i in range(1, m + 1):
                count += min(mid // i, n)
            if count < k:
                low = mid + 1
            else:
                high = mid
        return low
